### ISSUE TITLE
create and update workflow in data source details

### ISSUE THREAD

**Issue by josh-chamberlain** [Police-Data-Accessibility-Project/data-sources-app#163]
after #153 #18 #151 #150

**Comment by maxachis** [comment]
@josh-chamberlain Given that we will be hosting this within Retool, do any of the above requirements need changing? 

For example: 

> "for now, only admin users should be allowed to edit"
> non-auth'd users should not be able to see the edit button

Do we imagine anyone aside from admins will have access to the retool features? Or that we will have anyone other than non-authorized users?

**Comment by josh-chamberlain** [comment]
@maxachis this is tricky. Users should get a permission for database editing, and we could manually assign that permission somehow. Since we don't have proper version control right now, it's a bit of a scary permission to grant.

How about this? 

- for now: let's keep the retool username and password pretty much hidden
- sometime soon: give a short list of people the ability to edit the database as a permission; let people log into `retool` with their PDAP credentials via OAuth and let people look, but not touch (unless they have permission)
- also, maybe: generate a report using our logs of who edited the database; do this a day or two before they refresh. use discord for notifications.

**Comment by maxachis** [comment]
>  changes need to push back to Airtable for now, until we replace it, or it will sync back old data.
we'll need to flip a switch to make this start sending to our API; should we just do both, for now, redundantly? maybe we comment out one, then the other?

@josh-chamberlain This would be a challenging (but not impossible) task, for a few reasons:

1. The V2 API pulls data from the V2 Database, and both the API and the Database have been modified to account for various changes to the data sources schema (columns removed, some with revised types, et cetera). Thus, the data being edited in retool would need to be further modified so as to be inserted into Airtable. 
2. Testing proper functionality for the Retool App becomes considerably more high stakes when we're potentially inserting directly into the Airtable production database before confirming that everything is functioning correctly; with V2D, we can test with more comfort that any errors can be easily refreshed over.
3. Bear in mind that we have a script which syncs from Airtable to V1 Database on an hourly basis; if, while we're still beta testing the V2 Web App, we have a migration script which refreshes V1D -> V2D on an hourly or daily basis, we're facing a delay in terms of the Airtable data propagating to V2D, where it would then be ingested by the Retool app.
4. (Opinion) It strikes me as redundant; Airtable already has means for modifying its data. Creating a retool interface for interacting with Airtable, when we already have a means for interacting with Airtable doesn't strike me as adding much, even in the interim period.

As I said, it's not impossible, but it is challenging, and it would extend the time required to develop this. 

I think there are competing risks to having Retool -> Airtable & V2D vs. Retool -> V2D Alone -- the risks for the former I've outlined above, and the risks for the latter are, of course, what we're trying to account for _with_ the former: an abrupt shift, potential downtime and bugs in the transition, an interruption in service, etc. 

I don't want to discount those. And if this is determined to be the way to go, then so be it! But I think we should take into account the challenges I provided above and consider alternatives to the above for mitigating those risks.

**Comment by maxachis** [comment]
I've created a version of the Retool App. Still working on modifying to meet all requirements (aside from the one I'm discussing above).

https://pdapinternal.retool.com/apps/60948a44-9646-11ef-8664-ab94ecae4e51/Data%20Source%20Edit

**Comment by josh-chamberlain** [comment]
@maxachis you're right about syncing to Airtable—a bad idea! Sometimes I don't know what past me was thinking. I adjusted the issue, and added a note in the `context` section.

The bare minimum thing we need to do with a basic `search` is identify a source; worst case scenario, that looks like 

1. finding a source in the UI or our db
2. copying the UID
3. pasting it into retool, where the source is called up and editable

As long as Retool has some way to individually edit sources one by one by calling them up by UID, we're probably good. We can improve it later. Plus, we _do_ have the API, and we _do_ have the database, and in a pinch we can make something weird happen as necessary. This is just to hold us over until https://github.com/Police-Data-Accessibility-Project/data-sources-app/issues/477

**Comment by josh-chamberlain** [comment]
## Testing data sources edit

This is fantastic overall. The grouping, the selection based on data types...great work!

1. When I go to submit an edit, I consistently get this error...even when I think I'm adding all the required fields:

<img width="325" alt="Screen Shot 2024-11-04 at 3 27 58 PM" src="https://github.com/user-attachments/assets/a9f2b65f-e724-4a33-96ce-7fc03be68283">

2. Let's un-require `Access Types`
3. Let's move `tags` to Additional Record Info (it's seldom-used; we have so many properties that tags usually seem redundant, and I think their best use case is programmatic)
4. `Record Formats` should be a multi-select, much like Access Types

## Testing data request edit

Indeed, your location linker is wonderful! It makes much more sense than associating requests with agencies.

1. When I go to submit an edit, the form refreshes, clearing out the required fields I was trying to add.
2. Let's un-require `Record types required`, `Data requirements`, `Request urgency`, and `PDAP response`.
3. Let's make a group of properties called "Submission form" made up of [those found here](https://airtable.com/app473MWXVJVaD7Es/shrbFfWk6fjzGnNsk).
4. Others can be grouped into..."Management/organizing"?

**Comment by maxachis** [comment]
@josh-chamberlain Addressed all issues! Give it another whirl!

**Comment by maxachis** [comment]
> To make you feel better, I will set a lower bar: the _actual_ worst case scenario would be something like "every time you enter an ID that does not exist and try to load a single source, the database is wiped without a trace". You have done well to exceed that!

I wish to make it clear for the record that this made me laugh out loud.
